lib/vm/src/instance.rs

```diff
@@ -761,99 +680,98 @@ impl Instance {
     }
 }
 
-/// A handle holding an `Instance` of a WebAssembly module.
-#[derive(Hash, PartialEq, Eq)]
-pub struct InstanceHandle {
-    instance: *mut Instance,
-
-    /// Whether `Self` owns `self.instance`. It's not always the case;
-    /// e.g. when `Self` is built with `Self::from_vmctx`.
-    ///
-    /// This information is necessary to know whether `Self` should
-    /// deallocate `self.instance`.
-    owned_instance: bool,
+#[derive(Debug)]
+#[repr(C)]
+pub struct InstanceAllocator {
+    ptr_to_dealloc: *const Arc<Self>,
+    ptr_layout: Layout,
+    instance: mem::ManuallyDrop<Instance>,
 }
 
-/// # Safety
-/// This is safe because there is no thread-specific logic in `InstanceHandle`.
-/// TODO: this needs extra review
-unsafe impl Send for InstanceHandle {}
+impl InstanceAllocator {
+    #[inline(always)]
+    const fn new(instance: Instance, ptr_to_dealloc: *const Arc<Self>, ptr_layout: Layout) -> Self {
+        Self {
+            instance: mem::ManuallyDrop::new(instance),
+            ptr_to_dealloc,
+            ptr_layout,
+        }
+    }
 
-impl InstanceHandle {
-    /// Allocates an instance for use with `InstanceHandle::new`.
-    ///
-    /// Returns the instance pointer and the [`VMOffsets`] that describe the
-    /// memory buffer pointed to by the instance pointer.
-    pub fn allocate_instance(module: &ModuleInfo) -> (NonNull<u8>, VMOffsets) {
-        let offsets = VMOffsets::new(mem::size_of::<*const u8>() as u8, module);
+    fn instance_layout(offsets: &VMOffsets) -> Layout {
+        let size = mem::size_of::<Arc<mem::ManuallyDrop<Self>>>()
```

Comment:
This seems wrong to me 🤔

Reply:
But perhaps I don't really understand -- as far as I can tell the `Arc<ManuallyDrop<Self>>` is a lie

Reply:
It is. It's an error in my merge. Fixing it.